Prefer logging to console unless JOURNAL_STREAM is set

DaanDeMeyer opened the pull request:
Fixes #15902.

I tried adding a `log_setup_cli` function first in a similar vein to `log_setup_service` but there seem to be quite some discrepancies between all the tools (some set LC_ALL to "", some enable colors, etc). I'm not sure if we could align all those tools to use the exact same setup so I left that out for now.

keszybz commented on src/binfmt/binfmt.c:
```diff
@@ -191,7 +191,8 @@ static int run(int argc, char *argv[]) {
 
-        log_setup_service();
+        log_parse_environment();
+        log_open();
```
That's fairly ugly... What about going in the other direction: making log_setup_service have a parameter like log_to_console and putting `true` or `false` in the callers?

DaanDeMeyer commented:
There might be more offenders. I took the tools used in Arch pacman systemd hook and systemd-firstboot as those are the tools I've been impacted by the problem.

keszybz commented:
> but there seem to be quite some discrepancies between all the tools (some set LC_ALL to "", some enable colors, etc). 

The history here is that adding color output in logging requires some work when there was manual coloring added to messages. It is necessary to remove/adjust/verify that this coloring does not interact badly with the log coloring. That is why it's not turned on everywhere.

keszybz commented on src/basic/log.c:
```diff
@@ -1396,12 +1396,12 @@ int log_dup_console(void) {
          * terminal if invoked interactively. */
 
-        log_set_target(LOG_TARGET_AUTO);
+        log_set_target(log_to_console ? LOG_TARGET_CONSOLE : LOG_TARGET_AUTO);
```
Hmm, maybe we should combine this with an explicit setting in the units where the binaries that set log_to_console=true are used? We still want to log to the journal from systemd-tmpfiles.service, systemd-sysusers.service, etc.

DaanDeMeyer replied:
Is there a big difference between logging directly to the journal and having output pass via the console? It ends up in the journal regardless no?

DaanDeMeyer replied:
Setting the SYSTEMD_LOG_TARGET environment variable in the relevant units works but is a little bit ugly. Maybe there's a way to detect whether we're running as a service and switch between console and journal based on that?

keszybz replied:
Yes, there's a difference: when going via the console, all metadata is stripped (or rather, it is never attached in the first place). I agree that setting of a variable is ugly. Indeed, maybe it'd be better to look at $JOURNAL_STREAM? @poettering, opinions?

poettering commented:
i like @keszybz' idea with the JOURNAL_STREAM stuff. i.e. automatically upgrade to structured journal logging if stdout/stderr is actually connected to a journal stream anyway. It's precisely what $JOURNAL_STREAM was invented for. And this could easily be hidden inside log_setup_service() so that it magically works for all services

DaanDeMeyer commented:
We now prefer console unless `JOURNAL_STREAM` is set. This works but does make console preferred over kmsg by default. I'm not sure what the best way is to deal with that. Maybe its fine to prefer console over kmsg unless kmsg is explicitly configured?

poettering commented on src/basic/log.c:
```diff
@@ -1131,6 +1131,10 @@ void log_parse_environment_realm(LogRealm realm) {
                  * closed the console yet. */
                 (void) proc_cmdline_parse(parse_proc_cmdline_item, NULL, PROC_CMDLINE_STRIP_RD_PREFIX);
 
+        e = getenv("JOURNAL_STREAM");
```
so we set the env var to the st_dev and st_ino fields of the fd stderr is connected to. The idea is that this allows us to protect us from cases where we are forked off some service that itself had stderr connected to the journal but connects our stderr to a pipe it allocated itself, but didn#t clean up $JOURNAL_STREAM.

We should thus compare the contents with the data from fstat(STDERR_FILENO). i.e. add a new helper call or so:

```c
bool stderr_is_journal(void) {
        _cleanup_free_ char *w = NULL;
        const char *e;
        uint64_t dev, ino;

        e = getenv("JOURNAL_STREAM");
        if (!e)
                return false;

        if (extract_first_word(&e, &w, ":",  EXTRACT_DONT_COALESCE_SEPARATORS) <= 0)
                return false;

         if (safe_atou64(w, &dev) < 0)
                return false;
         if (safe_atou64(e, &ino) < 0)
                return false;        

        if (fstat(STDERR_FILENO, &st) < 0)
                return false;

        return st.st_dev == dev & st.st_ino == ino;
}
```

poettering replied:
i wonder if we should do this unconditionally in log_open() if LOG_TARGET_AUTO is chosen — entirely replacing the isatty() check

DaanDeMeyer replied:
I replaced the `isatty` check by the `stderr_is_journal` check. I think that's a net improvement over the `isatty` check.

Also, is `extract_many_words` supposed to succeed if it can't read the requested amount of words? That gave me a pretty nasty segfault.

poettering commented on src/basic/log.c:
```diff
@@ -219,6 +220,30 @@ static int log_open_journal(void) {
+        if (!e)
+                return false;
+
+        if (extract_many_words(&e, ":", EXTRACT_DONT_COALESCE_SEPARATORS, &f, &s, NULL) < 2)
```
hmm, to fully validate the input string you'd have to check `isempty(e)` after calling extract_many_words() to check that there's no trailing rubbish (think: a string line `47:11:8:15` is passed...)

I'd probably just use extract_first_word() to extract the first word, and then parse that to an integer and parse whatever remains as another integer (as my original code snippet suggested)

poettering commented:
I like this a lot I must say. This is a lot nicer than the old isatty() check.

poettering commented:
CI is unhappy?

mrc0mmand commented:
For some reason this breaks `TEST-21-SYSUSERS` in all CIs:

[CentOS CI - Arch](https://ci.centos.org/job/systemd-pr-build-vagrant/8026/artifact//systemd-centos-ci/artifacts_RQwWbQ/vagrant-logs.0WR/vagrant-arch-testsuite.YdZ/TEST-21-SYSUSERS_FAIL.log)
```
*** Running test unhappy-1.input
--- /var/tmp/systemd-test-TEST-21-SYSUSERS/tmp/err	2020-06-10 18:58:25.007480639 +0000
+++ unhappy-1.expected-err	2020-06-10 18:34:44.460643387 +0000
@@ -1 +1 @@
-Creating group tape with gid 968.
+Failed to parse UID: '9999999999': Numerical result out of range
**** Unexpected error output for unhappy-1.input
Creating group tape with gid 968.
make: *** [Makefile:4: run] Error 1
make: Leaving directory '/build/test/TEST-21-SYSUSERS'
```

[CentOS CI - CentOS 7](https://ci.centos.org/job/systemd-pr-build/12426/artifact//systemd-centos-ci/artifacts_XK4_t6/testsuite-logs-upstream.Ug4/TEST-21-SYSUSERS_FAIL.log)
```
*** Running test unhappy-1.input
--- /var/tmp/systemd-test-TEST-21-SYSUSERS/tmp/err	2020-06-10 19:38:21.061806943 +0100
+++ unhappy-1.expected-err	2020-06-10 19:25:33.560142414 +0100
@@ -1 +1 @@
-Creating user systemd-coredump (systemd Core Dumper) with uid 988 and gid 988.
+Failed to parse UID: '9999999999': Numerical result out of range
**** Unexpected error output for unhappy-1.input
Creating user systemd-coredump (systemd Core Dumper) with uid 988 and gid 988.
make: *** [run] Error 1
make: Leaving directory `/root/systemd/test/TEST-21-SYSUSERS'
```

Ubuntu CI - amd64
```
*** Running test unhappy-1.input
--- /var/tmp/systemd-test.6yW0xs/tmp/err	2020-06-11 04:03:06.964521988 +0000
+++ unhappy-1.expected-err	2020-06-11 02:18:14.000000000 +0000
@@ -0,0 +1 @@
+Failed to parse UID: '9999999999': Numerical result out of range
**** Unexpected error output for unhappy-1.input
make: *** [run] Error 1
Makefile:4: recipe for target 'run' failed
make: Leaving directory '/tmp/autopkgtest.c4GiLG/build.icl/systemd/test/TEST-21-SYSUSERS'
```

etc.

DaanDeMeyer commented:
Fixed, the tests relied on the output being logged to journald instead of the console. That gives us a good indication that the change is working. I fixed the tests to read the output from the console instead and they seem to be passing now.
